crates/oxc_transformer/src/es2020/nullish_coalescing_operator.rs

```diff
@@ -0,0 +1,125 @@
+    _ctx: Ctx<'a>,
+    var_declarations: std::vec::Vec<Vec<'a, VariableDeclarator<'a>>>,
+}
+
+impl<'a> NullishCoalescingOperator<'a> {
+    pub fn new(ctx: Ctx<'a>) -> Self {
+        Self { _ctx: ctx, var_declarations: vec![] }
+    }
+
+    pub fn transform_statements(
+        &mut self,
+        _statements: &mut Vec<'a, Statement<'a>>,
+        ctx: &mut TraverseCtx<'a>,
+    ) {
+        self.var_declarations.push(ctx.ast.vec());
+    }
+
+    pub fn transform_statements_on_exit(
+        &mut self,
+        statements: &mut Vec<'a, Statement<'a>>,
+        ctx: &mut TraverseCtx<'a>,
+    ) {
+        if let Some(declarations) = self.var_declarations.pop() {
+            if declarations.is_empty() {
+                return;
+            }
+            let variable = ctx.ast.alloc_variable_declaration(
+                SPAN,
+                VariableDeclarationKind::Var,
+                declarations,
+                false,
+            );
+            statements.insert(0, Statement::VariableDeclaration(variable));
+        }
+    }
+
+    fn create_new_var_with_expression(
+        &mut self,
+        expr: &Expression<'a>,
+        ctx: &mut TraverseCtx<'a>,
+    ) -> IdentifierReference<'a> {
+        // Add `var name` to scope
+        let name = match expr {
+            Expression::Identifier(ref ident) => ident.name.as_str(),
+            _ => "nullish_coalescing_operator",
+        };
+        let symbol_id =
+            ctx.generate_uid_in_current_scope(name, SymbolFlags::FunctionScopedVariable);
+        let symbol_name = ctx.ast.atom(ctx.symbols().get_name(symbol_id));
+
+        {
+            // var _name;
+            let binding_identifier = BindingIdentifier {
+                span: SPAN,
+                name: symbol_name.clone(),
+                symbol_id: Cell::new(Some(symbol_id)),
+            };
+            let kind = VariableDeclarationKind::Var;
+            let id = ctx.ast.binding_pattern_kind_from_binding_identifier(binding_identifier);
+            let id = ctx.ast.binding_pattern(id, None::<TSTypeAnnotation<'_>>, false);
+            self.var_declarations
+                .last_mut()
+                .unwrap()
+                .push(ctx.ast.variable_declarator(SPAN, kind, id, None, false));
+        };
+
+        ctx.create_reference_id(SPAN, symbol_name, Some(symbol_id), ReferenceFlag::Read)
+    }
+
+    pub fn transform_expression(&mut self, expr: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
+        // left ?? right
+        let Expression::LogicalExpression(logical_expr) = expr else { return };
+        if logical_expr.operator != LogicalOperator::Coalesce {
+            return;
+        }
+
+        let reference;
+        let assignment;
+
+        // skip creating extra reference when `left` is static
+        if ctx.symbols().is_static(&logical_expr.left) {
+            reference = ctx.ast.copy(&logical_expr.left);
+            assignment = ctx.ast.copy(&logical_expr.left);
```

Comment:
Please god no!

If you insert this after `if logical_expr.operator != LogicalOperator::Coalesce`, then you have an owned `LogicalExpression` and can remove all the `ast.copy` calls:

```rs
// Take ownership of the `LogicalExpression`
let logical_expr = match ctx.ast.move_expression(expr) {
    Expression::LogicalExpression(logical_expr) => logical_expr.unbox(),
    _ => unreachable!(),
};
```

This is pretty horrible, but should not be so bad performance-wise. The compiler should be able to deduce that it's a `LogicalExpression` (because we already checked that above) and remove the `unreachable!()` branch (because it is indeed provably unreachable).

As it stands, this code is definite UB, as you are using `logical_expr.left` and `reference` twice. Once you take out the `ast.copy` calls, the compiler draws your attention to this.

Reply:
Yes! You're right. I missed these `ctx.copy` usages. Now I applied your suggestion. That looks good!